torch/csrc/autograd/custom_function.h

```diff
@@ -342,7 +342,8 @@ variable_list CppNode<T>::apply(variable_list&& inputs) {
   variable_list backward_inputs;
   backward_inputs.reserve(num_inputs);
   for (const auto i : c10::irange(num_inputs)) {
-    if (inputs[i].defined() || !ctx_.materialize_grads_) {
+    if (inputs[i].defined() || !ctx_.materialize_grads_ ||
```

Comment:
Shouldn't these still be undefined even when materialize_grad is True ?

Reply:
Actually no, I don't think we want to do this now. 

Yes, I don't imagine anyone actually needing zero-filled grads for non-differentiable outputs, but I think this can break some people as Richard points out above - we also have quite a few failures in tests.

So for the purposes of making this minimally BC-breaking, I prefer to make this a opt-in mechanism instead, as in
I do want to materialize when ctx.set_materialize_grads(False).

My reasoning is that:
- it is rare for people who can't handle None to be using set_materialize_grads(False)
- people care about perf will find out about using set_materialize_grads(False)

Reply:
Sounds ok to me!